COMP: Deprioritize completion of &mut self methods for const references and bindings

Kobzol opened the pull request:
This PR attempts to sort methods in completion based on mutability of `self`. If a method call is performed on a const reference or a non-mut binding,`&mut self` methods will be deprioritized.

The implementation needs additional work that I wanted to discuss here. I needed to somehow pass the method call (`RsFieldLookup`) to `ScopedBaseCompletionEntity::getBasePriority` (I didn't know how else to get it from there). `ImplLookup` was already being passed down the completion chain, so I just slapped the element to the lookup, which is not optimal. Is there an easier way of passing down the element? Should I add it to the `RsCompletionContext` and pass it down the completion chain of methods and processors?

The implementation now uses a very simple heuristic for deciding if the method should be deprioritized. Let me know if I should improve it (or how :) ).

Completion could possibly receive a larger overhaul to support use cases like smart completion (https://github.com/intellij-rust/intellij-rust/issues/5196).

Fixes: https://github.com/intellij-rust/intellij-rust/issues/1409

mchernyavsky reviewed (commented):
> Should I add it to the `RsCompletionContext` and pass it down the completion chain of methods and processors?

Yes, you can pass the element to the completion context instead of `implLookup`:

```rust
data class RsCompletionContext(
    val context: RsElement? = null,
    val expectedTy: Ty? = null,
    val isSimplePath: Boolean = false
) {
    val lookup: ImplLookup? get() = context?.implLookup
}
```

Kobzol commented:
Ok, I moved it to `RsCompletionContext`.

I think that the lookup should be computed eagerly (as it was before) to avoid recomputing it on every access (i.e. I removed `get()`):
```kotlin
data class RsCompletionContext(
    val context: RsElement? = null,
    val expectedTy: Ty? = null,
    val isSimplePath: Boolean = false
) {
    val lookup: ImplLookup? = context?.implLookup
}
```

mchernyavsky reviewed (approved):
Thanks!

bors r+
